### PR TITLE
Refactor the users module to bring together all the functions for propagating user IDs.

### PR DESCRIPTION
**Pull request by tomwilkie**

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -64,18 +65,19 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
...
+	grpcListener net.Listener
+	httpServer   *http.Server
 
 	HTTP *mux.Router
 	GRPC *grpc.Server
 }
 
 // New makes a new Server
 func New(cfg Config) (*Server, error) {
+	// Setup listeners firt, so we can fail early if the port is in use.
```
Comment (juliusv) [inline server/server.go]: firt -> first

File: server/server.go [hunk 2]
```diff
@@ -86,39 +88,59 @@ func New(cfg Config) (*Server, error) {
 		return nil, err
 	}
 
+	// Prometheus historgrams for requests.
```
Comment (juliusv) [inline server/server.go]: historgrams -> histograms

**Comment by tomwilkie** [comment]
Thanks @juliusv!  I wrote this mainly for @bboreham, so I'll still wait for his response if thats okay...

File: server/server.go [hunk 3]
```diff
@@ -86,39 +88,59 @@ func New(cfg Config) (*Server, error) {
...
+		},
+		middleware.Instrument{
+			Duration:     requestDuration,
+			RouteMatcher: router,
+		},
+		middleware.Func(func(handler http.Handler) http.Handler {
+			return nethttp.Middleware(opentracing.GlobalTracer(), handler)
+		}),
+	}
+	httpMiddleware = append(httpMiddleware, cfg.HTTPMiddleware...)
```
Comment (tomwilkie) [inline server/server.go]: No no this isn't going to work, not all the http paths need the same middleware (ie auth)

Comment (tomwilkie) [inline server/server.go]: Actually its fine - usage of this library can add per-path middleware.

**Comment by tomwilkie** [comment]
CircleCI has passed - don't know why this isn't green.  I blame s3.

**Comment by tomwilkie** [comment]
@bboreham said:

> it's not particularly clear to me where one should call `ExtractFromHTTPRequest`

Thats very much up to the user - I see two options:
1) in your request handlers for requests that should be authenticated
2) by putting the http auth middleware somewhere in your mux: https://github.com/weaveworks/common/pull/22/commits/d5b134a763dd430baa9c982848a12f09fe3e0a68#diff-79c6c4f89ba3175bec7233e974e6dacf (this file might have been missing when you first looked)

File: user/http.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+	return userID, Inject(r.Context(), userID), nil
+}
+
+// InjectIntoHTTPRequest injects the userID from the context into the request headers.
+func InjectIntoHTTPRequest(ctx context.Context, r *http.Request) error {
+	userID, err := Extract(ctx)
+	if err != nil {
+		return err
+	}
+	r.Header.Add(orgIDHeaderName, userID)
```
Comment (bboreham) [inline user/http.go]: the analogous grpc code does several checks; should this code do them too?

Comment (tomwilkie) [inline user/http.go]: Yes, probably!

**Pull request merged**